caffe: use pre-CUDNN 8.x

ConnorBaker opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Since https://github.com/NixOS/nixpkgs/pull/215578, the default for cudatoolkit 10.1 is CUDNN 8.0.5, the last version to support CUDA 10.1. However, caffe does not build with CUDNN 8.x, so we use CUDNN 7.6.5 instead. Earlier versions of cudatoolkit use pre-8.x CUDNN, so we use the default.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ConnorBaker committed:
caffe: use pre-CUDNN 8.x

ConnorBaker commented:
Result of `nixpkgs-review pr 217330` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>caffeWithCuda</li>
  </ul>
</details>

ConnorBaker commented:
@abbradar @avitex would you mind giving this a quick once-over? The change is the addition of a conditional to ensure `caffe` isn't built with a version of CUDNN which is too new.

samuela commented:
cc @NixOS/cuda-maintainers

samuela commented:
Result of `nixpkgs-review pr 217330` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>caffeWithCuda</li>
  </ul>
</details>

samuela commented on pkgs/applications/science/math/caffe/default.nix:
```diff
@@ -21,11 +21,17 @@
 
 assert leveldbSupport -> (leveldb != null && snappy != null);
-assert cudnnSupport -> cudaSupport;
+assert cudnnSupport -> cudaSupport && lib.versionOlder cudnn.version "8.0";
```
given this assert, is there any situation in which this package would work with cudnn 7.6.5 as is selected in the let statement above? 

why not just drop support for older cuda/cudnn versions for this package?

ConnorBaker replied:
You're right, the assert there is redundant, I'll remove it.

Per your second question, this package *only* works with older versions of CUDA/CUDNN. Since it won't work with CUDNN 8.x+, it can't run on Ampere or newer.

ConnorBaker commented:
Removed the useless check in assert that CUDNN was of the correct version given that the `let` statement handles that.

Also removed the PR stack info in the header because it can be merged independently of the others.

samuela commented:
Is there a reason to support cuda < 10.1? Why not just mark caffe as broken for cuda versions that are too old?

ConnorBaker commented:
Maybe the way I wrote the `if` expression is weird -- CUDA 10.0 and all the prior versions of CUDA work fine because they only support *up to* CUDNN 7.6.5. (That's why we use their default `cudaPackages.cudnn` and don't need to fish around for a different version.)

CUDA 10.1 is special because it's the only release in the 10.x series to be supported by a version of CUDNN 8 (it's supported by CUDNN 8.0.5). So in the special case where we have 10.1, we need to fall back to the older version of CUDNN.

samuela commented:
Ahh, I misunderstood -- thank you for explaining! Ok this LGTM then

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-45/26397/1
